### DISCUSSION TITLE
Unable to remove listener for relay

### DISCUSSION THREAD

**Discussion by dariusc93** [libp2p/rust-libp2p#4468] [Q&A]
Not sure if this is by designed, but in the current implementation, one might expect to be able to stop listening on the circuit relay by simply calling `Swarm::remove_listener`, or emit `ToSwarm::RemoveListener` from the behaviour, however supplying the listener id of the active relay thats being listened on by the local node to one of those calls does not actually stop the node from using the relay nor does it emit an event to `Swarm` (nor does the log indicate anything being done). I may be missing a piece of info but the relay client transport [does allow closing the listener](https://github.com/libp2p/rust-libp2p/blob/c0bccf724e91a7adb6cb2a0015329691b46f2030/protocols/relay/src/priv_client/transport.rs#L157-L164), however this doesnt seem to be the case when actually trying to close it. To be able to stop listening on the relay, one would have to close the connection, however, this may not be desirable if the node wishes to maintain connection without having to disconnect then reconnect.

I bought this up in https://github.com/libp2p/rust-libp2p/discussions/3430 but forgot about it (including forgetting to follow up), however this look more of a bug than being intentional by design, but looking over the spec, it does make one believe the only way is to actually disconnect. Maybe somebody can confirm this if possible

**Comment by dariusc93** [comment]
/CC @mxinden any possible insight on this matter?

**Comment by thomaseizinger** [comment]
To clarify, you'd like to have a connection to the relay without an active reservation?

I've not thought deeply about it but at first sight, having `remove_listener` do something meaningful sounds like a good idea.

**Reply by dariusc93** [reply]
I would like to be able to close the relay (or reservation) without closing the connection to the relay node itself, if that make sense. Pretty much wanting to preserve the connection while cancelling the reservation (by removing the listener). 

As of right now, I am having to close the connection to the relay, which isnt exactly a solution I wanted to do but does work to get it done for now.

**Reply by thomaseizinger** [reply]
I see. I think that is very reasonable. The problem will be that we don't have a message to tell the relay that we no longer want the reservation :)

See https://github.com/libp2p/specs/blob/master/relay/circuit-v2.md.

So I think the only thing we can do is have the current reservation expire which would mean that for the remaining period, other nodes could still connect to you. Would that help your usecase?

**Reply by dariusc93** [reply]
Sorry for the late response. That should work fine for my use case.

**Reply by thomaseizinger** [reply]
> Sorry for the late response. That should work fine for my use case.

Cool. If I understand the current code correctly, than this already happens. As soon as the listener is closed, the channel between behaviour and transport gets de-allocated and we de-allocate our reserveration as a result:

https://github.com/libp2p/rust-libp2p/blob/1fff308d19a614a6356ff446b5507b56b0d0ddee/protocols/relay/src/priv_client/handler.rs#L638-L641
